crypto/ec/ec_key.c

```diff
@@ -267,6 +267,56 @@ int EC_KEY_check_key(const EC_KEY *eckey)
     return eckey->group->meth->keycheck(eckey);
 }
 
+/*
+ * Check the range of the EC public key.
+ * See SP800-56A R3 Section 5.6.2.3.3 (Part 2)
+ * i.e.
+ *  - If q = odd prime p: Verify that xQ and yQ are integers in the
+ *    interval[0, p − 1], OR
+ *  - If q = 2m: Verify that xQ and yQ are bit strings of length m bits.
+ */
+static int ec_key_public_range_check(BN_CTX *ctx, const EC_KEY *key)
+{
+    int ret = 0;
+    BIGNUM *x, *y;
+
+    BN_CTX_start(ctx);
+    x = BN_CTX_get(ctx);
+    y = BN_CTX_get(ctx);
+    if (y == NULL)
+        goto err;
+
+    if (!EC_POINT_get_affine_coordinates(key->group, key->pub_key, x, y, ctx))
+        goto err;
+
+    if (EC_METHOD_get_field_type(key->group->meth) == NID_X9_62_prime_field) {
+        if (BN_is_negative(x)
+            || BN_cmp(x, key->group->field) >= 0
+            || BN_is_negative(y)
+            || BN_cmp(y, key->group->field) >= 0) {
+            goto err;
+        }
+    } else {
+        int m = EC_GROUP_get_degree(key->group);
+        if (BN_num_bits(x) != m || BN_num_bits(y) != m) {
```

Comment:
This looks off. Yes, a reduced element of GF(2<sup>m</sup>) should be an m-bit string (i.e. every coefficient beyond x<sup>m-1</sup> is zero), but OpenSSL's internal representation is not a bit string. OpenSSL stuffs it into a `BIGNUM` but implements polynomial operations instead of integer operations (see `BN_GF2m_add`). `BN_num_bits` returns the number of bits it takes to *minimally* represent the absolute value of the `BIGNUM` as an integer or, in polynomial terms, one plus the highest non-zero power. This code will incorrectly reject field elements that happen to have the x<sup>m-1</sup> coefficient as zero, which is not what "bit strings of length m bits" means.

This check presumably should align with, say, this logic here:
https://github.com/openssl/openssl/blob/master/crypto/ec/ec2_oct.c#L296-L301

Although that too also appears to be wrong. `group->field`, for GF(2<sup>m</sup>) appears to be to the irreducible polynomial which defines the field, which has terms below x<sup>m</sup>. For instance, sect163k1 is defined by the polynomial x<sup>163</sup>+x<sup>7</sup>+x<sup>6</sup>+x<sup>3</sup>+1. (You take GF(2)[x] and mod out by that polynomial.) The code linked above will incorrectly accept, e.g., x<sup>163</sup>+1 and x<sup>163</sup>+x for being "less than" x<sup>163</sup>+x<sup>7</sup>+x<sup>6</sup>+x<sup>3</sup>+1 when interpreted as integers in binary. I place "less than" in quotes because that is not really a meaningful operation in this domain.

Reply:
For some strange reason I re-read the above comment about bit strings of length m bits, and then changed the > to a != without thinking about the side effects,
I think you are correct as it doesnt look like != will work. If I change it to > is that closer?
Here are a few public key x,y pairs currently being output for the 163k1 case..
p = 08:00:00:00:00:00:00:00:00::00::00:00:00:00:00:00:00:00:00:00:C9 (this is 164 bits)
m = 163 (164-1)
x,y pairs generated:
01:cc:1d:f3:aa:43:c9:70:dd:4c:18:73:ba:3f:96:93:60:c6:02:02:1f:
06:23:db:7c:2a:f9:8b:ca:1f:d8:78:9d:8d:90:40:24:7c:a8:a7:08:68

02:1a:76:10:21:6d:d6:ab:da:a3:a7:33:04:63:68:67:20:40:4c:d2:2f:
02:a6:ed:cf:99:f2:9f:a0:79:b8:b1:c2:2a:09:e9:ca:a9:0b:4b:1c:7d

06:b9:5b:40:cd:ae:fb:fa:62:27:35:19:87:1b:d5:c1:7b:2d:7b:95:1e:
01:30:b1:ac:f9:c7:f3:df:00:5a:e6:f1:3a:dc:e3:19:ca:9b:62:0c:94

05:cf:1c:23:c3:cc:f4:7d:e5:04:f4:8d:0f:32:b9:2f:c0:2f:0a:8b:e5:
04:ce:e7:be:f6:91:d1:86:f7:e6:45:fb:74:df:8b:16:93:57:01:a0:02

05:cc:73:5c:93:de:ff:8b:20:fc:df:6e:e0:e0:c9:81:4c:03:79:55:e3:
00:63:4e:72:da:8d:2b:15:a3:69:5d:39:3c:03:d8:92:bf:6b:79:2b:92

07:aa:75:c2:9c:11:46:e7:0c:81:0e:4b:55:7e:34:04:47:24:50:c1:72:
07:b8:3f:1e:38:74:83:fe:0a:8a:8c:26:31:22:d8:eb:1e:ab:f6:bf:a3

Reply:
Good catch, @davidben . @slontis if bugs like this are slipping by, IMO it means you need unit tests.

Reply:
I believe using > instead of != will give the right output, yeah. It's kinda goofy, but the goofiness comes from internal representation problems, which is a deeper issue (#6640).

ec2_oct.c is still wrong, but perhaps that should be fixed separately so it can be backported. Accidentally allowing unreduced inputs from the outside world probably isn't a great plan.

Reply:
> Edit: If you're not sure how to do that. If you XOR in the field polynomial to those (x,y) pairs, you'll get something that is still on the curve but your validation should reject. (In mod p terms, this is like adding p to x and y.)

Thanks

Reply:
test added

Reply:
unit test added for this case

Reply:
The fix to https://github.com/openssl/openssl/pull/8564#discussion_r269027104 has been put into a separate PR https://github.com/openssl/openssl/pull/8607. Thanks @davidben for pointing it out.